### PR TITLE
remove builtin use statement when lowering (towards #1618)

### PR DESCRIPTION
**Pull request by rupertford**
We add a wildcard use statement to the algorithm layer so that the PSyIR does not complain about builtin symbols not being declared. However, we also need to remove this wildcard use statement when lowering as the use statement does not really exist and will cause linking problems when compiling. This PR removes any use statements that were previously added by PSyclone.

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for first review from @arporter or @sergisiso

**Comment by rupertford** [comment]
Taking this back to avoid possible name clashes

**Comment by rupertford** [comment]
Ready for review again.

File: src/psyclone/generator.py [hunk 1]
```diff
@@ -581,23 +592,24 @@ def check_psyir(psyir, filename):
...
 
-def add_builtins_use(fp2_tree):
-    '''Modify the fparser2 tree adding a 'use builtins' so that builtin kernel
+def add_builtins_use(fp2_tree, name):
+    '''Modify the fparser2 tree adding a 'use <name>' so that builtin kernel
     functors do not appear to be undeclared.
 
     :param fp2_tree: the fparser2 tree to modify.
     :type fp2_tree: py:class:`fparser.two.Program`
+    :param str name: the name of the use statement.
```
Comment (arporter) [inline src/psyclone/generator.py:601]: It's picky but "...name of the module imported by the user statement."

File: src/psyclone/generator.py [hunk 2]
```diff
@@ -349,6 +349,17 @@ def generate(filename, api="", kernel_paths=None, script_name=None,
...
             invoke_trans.apply(
                 invoke, options={"kernels": kernels[id(invoke)]})
+        if api == "dynamo0.3":
+            # Remove any use statements that were temporarily added to
+            # avoid the PSyIR complaining about undeclared builtin
+            # names.
+            for node in psyir.walk((Routine, Container)):
+                symbol_table = node.symbol_table
+                try:
+                    symbol = symbol_table.lookup(builtins_module_name)
```
Comment (arporter) [inline src/psyclone/generator.py:359]: You could do `if builtins_module_name in symbol_table:` here and then you don't need the try..except?

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good but I think we could probably tighten things up a bit more.
Examples are fine with compilation.
Integration tests are all looking good but I'll check them at the next review (https://github.com/stfc/PSyclone-mirror/actions).
No need to update docs.
This is a 'towards' PR.

File: src/psyclone/generator.py [hunk 3]
```diff
@@ -276,10 +276,10 @@ def generate(filename, api="", kernel_paths=None, script_name=None,
         reader = FortranReader()
         if api == "dynamo0.3":
             # avoid undeclared builtin errors in PSyIR by adding "use
-            # builtins". TODO issue #1618. This symbol needs to be
-            # removed when lowering.
+            # psyclone_builtins".
+            builtins_module_name = "psyclone_builtins"
             fp2_tree = parse_fp2(filename)
-            add_builtins_use(fp2_tree)
+            add_builtins_use(fp2_tree, builtins_module_name)
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/generator.py:282]: I've made it an invalid Fortran name by using an underscore. I had to work around the fparser2 checks to implement this.

File: src/psyclone/generator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/generator.py:359]: I think we are OK now given that the name is invalid Fortran.

File: src/psyclone/generator.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/generator.py:601]: Modified.

File: src/psyclone/generator.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/generator.py:359]: OK, although we now lookup the symbol twice.

File: src/psyclone/tests/generator_test.py [hunk 4]
```diff
@@ -786,6 +786,30 @@ def test_generate_trans_error(tmpdir, capsys, monkeypatch):
...
+        "    use field_mod, only : field_type\n"
+        "    type(field_type) :: field\n"
+        "    call invoke(setval_c(field, 0.0))\n"
+        "  end subroutine test\n"
+        "end module\n")
+    filename = str(tmpdir.join("alg.f90"))
+    with open(filename, "w", encoding='utf-8') as my_file:
+        my_file.write(code)
+    alg, _ = generate(filename, api="dynamo0.3")
+    assert "use psyclone_builtins" not in alg
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:810]: No, the LFRic compilation testing support is limited to the PSy layer.

File: src/psyclone/tests/generator_test.py [hunk 5]
```diff
@@ -1176,9 +1200,8 @@ def test_no_script_lfric_new(monkeypatch):
     assert " testkern_type" not in alg
     # module symbol is removed
     assert "testkern_mod" not in alg
-    # TODO issue #1618. The builtins statement should be removed from
-    # the processed source code.
-    assert "use builtins" in alg
+    # builtins symbol (that was added by PSyclone) is removed
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:1203]: Oops.

File: src/psyclone/tests/generator_test.py [hunk 6]
```diff
@@ -1232,9 +1254,8 @@ def test_builtins_lfric_new(monkeypatch):
     assert " testkern_mod" not in alg
     assert " testkern_wtheta_mod" not in alg
     assert " testkern_w2_only_mod" not in alg
-    # TODO issue #1618. The builtins statement should be removed from
-    # the processed source code.
-    assert "use builtins" in alg
+    # builtins symbol (that was added by PSyclone) is removed
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/generator_test.py:1257]: Oops.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Tests, coverage and integration tests are all green.
No doc updates required.
Changes are pylint clean.
Will proceed to merge.

**Pull request merged**